### ISSUE TITLE
New rule for object literal keys ordering

### ISSUE THREAD

**Issue by arusakov** [palantir/tslint#1670]
There is only one rule for it: "object-literal-sort-keys". This is rule about alphabetical order.

I have another idea about keys ordering in TypeScript. In TypeScript we have `types` and `intarfaces` for describe object and for keys order. Look at this:
```ts
type X = {
  type: string;
  id: number;
} // describe my order (not alphabetical)

const x: X = {
  type: "type1",
  id: 1,
}; // it's good, because follows X keys order

const y: X = {
  id: 2,
  type: "type2",
}; // it's bad, because doesn't follow right order
```
I wrote sketch of this rule by myself [tslint-object-literal-smart-keys](https://github.com/arusakov/tslint-object-literal-smart-keys) with some [tests](https://github.com/arusakov/tslint-object-literal-smart-keys/tree/master/tests). It supports checking for object literal as argument, as return and as variable declarations. Nesting and optional properties are supported too.
What do you think about it?

**Comment by JoshuaKGoldberg** [comment]
@arusakov interesting proposal. We could add this as an option to `object-literal-sort-keys` by turning it into an optionally typed rule and requiring type information for the option... but that seems quite expensive. Let's hear if there's a strong need for it?

FWIW I'm -1 in general on these types of alphabetization rules. They tend to not be well received by many developers.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
